boards/shared/Src/Io/Io_SharedFrequencyOnlyPwmInput.c

```diff
@@ -0,0 +1,127 @@
+#include <main.h>
+#include <math.h>
+#include <assert.h>
+#include "Io_SharedFrequencyOnlyPwmInput.h"
+
+#define MAX_TIMER_PERIOD_ELAPSED_COUNT 2
+#define AUTO_RELOAD_REG 65535
+
+struct FreqOnlyPwmInput
+{
+    float              frequency_hz;
+    TIM_HandleTypeDef *htim;
+    float              timer_frequency_hz;
+    uint32_t           rising_edge_tim_channel;
+    bool               first_tick;
+    uint32_t           curr_rising_edge;
+    uint32_t           prev_rising_edge;
+};
+
+/**
+ * Set the frequency for the given PWM input
+ * @param pwm_input: The PWM input to set frequency for
+ * @param frequency_hz: Frequency, in Hz
+ */
+static void Io_SetFrequency(
+    struct FreqOnlyPwmInput *pwm_input,
+    const float              frequency_hz);
+
+static void Io_SetFrequency(
+    struct FreqOnlyPwmInput *const pwm_input,
+    const float                    frequency_hz)
+{
+    assert(frequency_hz >= 0.0f || isnanf(frequency_hz));
+
+    pwm_input->frequency_hz = frequency_hz;
+}
+
+struct FreqOnlyPwmInput *Io_SharedFreqOnlyPwmInput_Create(
+    TIM_HandleTypeDef *htim,
+    float              timer_frequency_hz,
+    uint32_t           rising_edge_timer_channel)
+{
+    assert(htim != NULL);
+    static struct FreqOnlyPwmInput pwm_inputs[MAX_NUM_OF_FREQ_ONLY_PWM_INPUTS];
+    static size_t                  alloc_index = 0;
+
+    assert(alloc_index < MAX_NUM_OF_FREQ_ONLY_PWM_INPUTS);
+
+    struct FreqOnlyPwmInput *const pwm_input = &pwm_inputs[alloc_index++];
+
+    pwm_input->htim                    = htim;
+    pwm_input->timer_frequency_hz      = timer_frequency_hz;
+    pwm_input->rising_edge_tim_channel = rising_edge_timer_channel;
```

Comment:
We should initialize `prev_rising_edge` and `curr_rising_edge` to known values, even if it's just 0.

Reply:
fixed